elements/lisk-chain/src/data_access/storage.ts

```diff
@@ -12,291 +12,537 @@
-			// eslint-disable-next-line camelcase
-			{ height_gte: fromHeight, height_lte: toHeight },
-			{ limit: null, sort: 'height:desc' },
-		);
+	): Promise<BlockHeaderJSON[]> {
+		const stream = this._db.createReadStream({
+			gte: `${DB_KEY_BLOCKS_HEIGHT}${formatInt(fromHeight)}`,
+			lte: `${DB_KEY_BLOCKS_HEIGHT}${formatInt(toHeight)}`,
+			reverse: true,
+		});
+		const blockIDs = await new Promise<string[]>((resolve, reject) => {
+			const ids: string[] = [];
+			stream
+				.on('data', ({ value }) => {
+					ids.push(value);
+				})
+				.on('error', error => {
+					reject(error);
+				})
+				.on('end', () => {
+					resolve(ids);
+				});
+		});
 
-		return blocks;
+		return this.getBlockHeadersByIDs(blockIDs);
 	}
 
 	public async getBlockHeadersWithHeights(
 		heightList: ReadonlyArray<number>,
-	): Promise<BlockJSON[]> {
-		const blocks = await this._storage.entities.Block.get(
-			{
-				// eslint-disable-next-line camelcase
-				height_in: heightList,
-			},
-			{
-				sort: 'height:asc',
-				limit: heightList.length,
-			},
-		);
-
+	): Promise<BlockHeaderJSON[]> {
+		const blocks = [];
+		for (const height of heightList) {
+			const block = await this.getBlockHeaderByHeight(height);
+			blocks.push(block);
+		}
 		return blocks;
 	}
 
-	public async getLastBlockHeader(): Promise<BlockJSON> {
-		const [lastBlockHeader] = await this._storage.entities.Block.get(
-			{},
-			{ limit: 1, sort: 'height:desc' },
-		);
+	public async getLastBlockHeader(): Promise<BlockHeaderJSON> {
+		const stream = this._db.createReadStream({
+			gte: getFirstPrefix(DB_KEY_BLOCKS_HEIGHT),
+			lte: getLastPrefix(DB_KEY_BLOCKS_HEIGHT),
+			reverse: true,
+			limit: 1,
+		});
+		const [blockID] = await new Promise<string[]>((resolve, reject) => {
+			const ids: string[] = [];
+			stream
+				.on('data', ({ value }) => {
+					ids.push(value);
+				})
+				.on('error', error => {
+					reject(error);
+				})
+				.on('end', () => {
+					resolve(ids);
+				});
+		});
 
-		return lastBlockHeader;
+		return this.getBlockHeaderByID(blockID);
 	}
 
 	public async getLastCommonBlockHeader(
 		arrayOfBlockIds: ReadonlyArray<string>,
-	): Promise<BlockJSON> {
-		const [block] = await this._storage.entities.Block.get(
-			{
-				// eslint-disable-next-line camelcase
-				id_in: arrayOfBlockIds,
-			},
-			{ sort: 'height:desc', limit: 1 },
-		);
-
-		return block;
+	): Promise<BlockHeaderJSON> {
+		const blocks = [];
+		for (const id of arrayOfBlockIds) {
+			try {
+				const block = await this.getBlockHeaderByID(id);
+				blocks.push(block);
+			} catch (error) {
+				if (!(error instanceof NotFoundError)) {
+					throw error;
+				}
+			}
+		}
+		blocks.sort((a, b) => b.height - a.height);
+
+		return blocks[0];
 	}
 
 	public async getBlocksCount(): Promise<number> {
-		const count = await this._storage.entities.Block.count({}, {});
-
-		return count;
+		const lastBlock = await this.getLastBlockHeader();
+		return lastBlock.height;
```

Comment:
This will be a behavioral change from the previous implementation. If the genesis block has a particular height in case of re-genesis then both of the implementations will return differen values. Do we really want that change?

Reply:
that's true, I think we don't have a way to get the total number of blocks other than iterating through the keys.
However, this will be removed in https://github.com/LiskHQ/lisk-sdk/issues/5320